File: code/legacy_server/utils/Logger.cs
// Copyright (C) NOMAD Group <nomad-group.net>.
// For licensing information see LICENSE at the root of this distribution.

using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace noda
{

    public static class Logger
    {
        private static List<ILogSink> sinks;

        public static void Init()
        {
            sinks = new List<ILogSink>();
        }

        public static void AddSink(ILogSink sink)
        {
            Debug.Assert(sink != null);

            sinks.Add(sink);
        }

        public static void Error(string text)
        {
            Log(LogLevel.Error, text);
        }

        public static void Critical(string text)
        {
            Log(LogLevel.Critical, text);
        }

        public static void Info(string text)
        {
            Log(LogLevel.Info, text);
        }

        public static void Trace(string text)
        {
            Log(LogLevel.Trace, text);
        }

        public static void Warning(string text)
        {
            Log(LogLevel.Warning, text);
        }

        public static void Log(LogLevel level, string text)
        {
            foreach (ILogSink sink in sinks)
            {
                sink.Write("", text, level);
            }
        }
    }
}

File: code/sync_server/server/utils/Logger.cpp
// Copyright (C) NOMAD Group <nomad-group.net>.
// For licensing information see LICENSE at the root of this distribution.

#include "Logger.h"

namespace noda::utils {

  static logcallback_t s_callback{ nullptr };

  static const char *GetLevelName(LogLevel log_level)
  {
#define LVL(x)      \
  case LogLevel::x: \
	return #x
	switch(log_level) {
	  LVL(Trace);
	  LVL(Info);
	  LVL(Warning);
	  LVL(Error);
	default:
	  return "???";
	}
#undef LVL
  }

  // pretty print me, please :-)
  void PrintLogMessageImpl(LogLevel ll, const char *text, const fmt::format_args &args)
  {
	const char *levelName = GetLevelName(ll);
	auto fmt = fmt::format("[{}]: {}", levelName, fmt::vformat(text, args));

	if(s_callback)
	  s_callback(fmt.c_str());
  }

  void SetLogCallback(logcallback_t callback)
  {
	s_callback = callback;
  }
} // namespace noda::utils